crates/assists/src/utils/insert_use.rs

```diff
@@ -86,21 +82,56 @@ impl ImportScope {
     }
 }
 
+pub(crate) enum InsertUseResult {
+    /// An import syntax node has to be inserted
+    Insert {
+        insert_pos: InsertPosition<SyntaxElement>,
+        elements_to_be_inserted: Vec<SyntaxElement>,
+    },
+    /// An import syntax node has to be replaced
+    Replace { node_to_be_replaced: SyntaxNode, node_to_be_inserted: SyntaxNode },
+}
+
+impl InsertUseResult {
+    pub(crate) fn apply(
+        self,
+        scope: &ImportScope,
+        builder: &mut crate::assist_context::AssistBuilder,
+    ) {
+        match self {
+            InsertUseResult::Insert { insert_pos, elements_to_be_inserted } => builder.insert(
+                match insert_pos {
+                    InsertPosition::First => TextSize::from(0),
+                    InsertPosition::Last => {
+                        TextSize::from(scope.as_syntax_node().text_range().end())
+                    }
+                    InsertPosition::Before(anchor) => anchor.text_range().start(),
+                    InsertPosition::After(anchor) => anchor.text_range().end(),
+                },
+                elements_to_be_inserted.into_iter().format("").to_string(),
+            ),
+            InsertUseResult::Replace { node_to_be_replaced, node_to_be_inserted } => {
+                builder.replace(node_to_be_replaced.text_range(), node_to_be_inserted.to_string())
```

Comment:
What if we cut of the offset determination logic out of `apply` and remove this method entirely?
The idea is to provide the data for both cases and simply require the user to match the enum data and apply this data himself.

This way we won't depend on the builder and provide enough data to call the corresponding builder method in a couple of lines only.

Reply:
This also seems like a decent option I suppose? I will give both a try, this and returning SyntaxWriter